middleware.go

```diff
@@ -0,0 +1,640 @@
+		if err != nil {
+			return fmt.Errorf("failed to decode payload.after: %w", err)
+		}
+		rec.Payload.After = decodedPayloadAfter
+	}
+
+	return nil
+}
+
+func (p *processorWithSchemaDecode) decode(ctx context.Context, data opencdc.Data, subject string, version int) (opencdc.StructuredData, error) {
+	switch data := data.(type) {
+	case opencdc.StructuredData:
+		return data, nil // already decoded
+	case opencdc.RawData: // let's decode it
+	default:
+		return nil, fmt.Errorf("unexpected data type %T", data)
+	}
+
+	sch, err := sdkschema.Get(ctx, subject, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get schema for %s:%d: %w", subject, version, err)
+	}
+
+	var structuredData opencdc.StructuredData
+	err = sch.Unmarshal(data.Bytes(), &structuredData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal bytes with schema: %w", err)
+	}
+
+	return structuredData, nil
+}
+
+// -- ProcessorWithSchemaEncode ----------------------------------------------
+
+const (
+	configProcessorSchemaEncodePayloadEnabled = "sdk.schema.encode.payload.enabled"
+	configProcessorSchemaEncodeKeyEnabled     = "sdk.schema.encode.key.enabled"
+)
+
+// ProcessorWithSchemaEncodeConfig is the configuration for the
+// ProcessorWithSchemaEncode middleware. Fields set to their zero value are
+// ignored and will be set to the default value.
+//
+// ProcessorWithSchemaEncodeConfig can be used as a ProcessorMiddlewareOption.
+type ProcessorWithSchemaEncodeConfig struct {
+	// Whether to encode the record payload with a schema.
+	// If unset, defaults to true.
+	PayloadEnabled *bool
+	// Whether to encode the record key with a schema.
+	// If unset, defaults to true.
+	KeyEnabled *bool
+}
+
+// Apply sets the default configuration for the ProcessorWithSchemaEncode middleware.
+func (c ProcessorWithSchemaEncodeConfig) Apply(m ProcessorMiddleware) {
+	if s, ok := m.(*ProcessorWithSchemaEncode); ok {
+		s.Config = c
+	}
+}
+
+func (c ProcessorWithSchemaEncodeConfig) SchemaPayloadEnabledParameterName() string {
+	return configProcessorSchemaEncodePayloadEnabled
+}
+
+func (c ProcessorWithSchemaEncodeConfig) SchemaKeyEnabledParameterName() string {
+	return configProcessorSchemaEncodeKeyEnabled
+}
+
+func (c ProcessorWithSchemaEncodeConfig) parameters() config.Parameters {
+	return config.Parameters{
+		configProcessorSchemaEncodePayloadEnabled: {
+			Default:     strconv.FormatBool(*c.PayloadEnabled),
+			Type:        config.ParameterTypeBool,
+			Description: "Whether to encode the record payload with a schema.",
+		},
+		configProcessorSchemaEncodeKeyEnabled: {
+			Default:     strconv.FormatBool(*c.KeyEnabled),
+			Type:        config.ParameterTypeBool,
+			Description: "Whether to encode the record key with a schema.",
+		},
+	}
+}
+
+// ProcessorWithSchemaEncode is a middleware that encodes the record
+// payload and key with a schema. It only encodes the record key/payload if the
```

Comment:
Tiny nitpick: payload and key => payload and/or key (to make sure users understand that only one of them can be configured).